Detectors/TPC/reconstruction/src/HwClusterer.cxx

```diff
@@ -8,461 +8,530 @@
+void HwClusterer::writeOutputWithTimeOffset(int timeOffset)
+{
+  // Check in which regions cluster were found
+  for (unsigned short region = 0; region < 10; ++region) {
+    if (mTmpClusterArray[region]->size() == 0)
+      continue;
 
-  mLastMcDigitTruth.erase(eventCount-mTimebinsPerCF);
+    if (mClusterArray) {
+      // Create new container
+      mClusterArray->emplace_back();
+      auto clusterContainer = mClusterArray->back().getContainer();
+
+      // Set meta data
+      clusterContainer->CRU = mClusterSector * 10 + region;
+      clusterContainer->numberOfClusters = 0;
+      clusterContainer->timeBinOffset = timeOffset;
+
+      for (auto& c : *mTmpClusterArray[region]) {
+        // if the container is full, create a new one
+        if (clusterContainer->numberOfClusters == mClusterArray->back().getMaxNumberOfClusters()) {
+          mClusterArray->emplace_back();
+          clusterContainer = mClusterArray->back().getContainer();
+          clusterContainer->CRU = mClusterSector * 10 + region;
+          clusterContainer->numberOfClusters = 0;
+          clusterContainer->timeBinOffset = timeOffset;
+        }
+        // Copy cluster and increment cluster counter
+        clusterContainer->clusters[clusterContainer->numberOfClusters++] = c.first;
+        for (auto& mcLabel : c.second) {
+          mClusterMcLabelArray->addElement(mClusterCounter, mcLabel.first);
+        }
+        ++mClusterCounter;
+      }
+    } else if (mPlainClusterArray) {
+      short cru = mClusterSector * 10 + region;
+      for (auto& c : *mTmpClusterArray[region]) {
+        auto& cluster = c.first;
+        mPlainClusterArray->emplace_back(
+          cru,
+          cluster.getRow(),
+          cluster.getQTotFloat(),
+          cluster.getQMax(),
+          cluster.getPad(),
+          std::sqrt(cluster.getSigmaPad2()),
+          cluster.getTimeLocal() + timeOffset,
+          std::sqrt(cluster.getSigmaTime2()));
+
+        for (auto& mcLabel : c.second) {
+          mClusterMcLabelArray->addElement(mClusterCounter, mcLabel.first);
+        }
+        ++mClusterCounter;
+      }
+    }
 
-  LOG(DEBUG) << "Event ranged from time bin " << iTimeBinMin << " to " << iTimeBinMax << "." << FairLogger::endl;
+    // Clear copied temporary storage
+    mTmpClusterArray[region]->clear();
+  }
 }
 
-void HwClusterer::ProcessTimeBins(int iTimeBinMin, int iTimeBinMax)
+//______________________________________________________________________________
+void HwClusterer::findClusterForTime(int timebin)
 {
+  if (timebin < 0)
+    return;
+
+  ClusterHardware cluster;
+  std::vector<std::pair<MCCompLabel, unsigned>> mcLabels;
+
+  unsigned qMaxIndex;
+  for (unsigned short row = mNumRows; row--;) {
+    // two empty pads on the left and right without a cluster peak
+    for (short pad = 2; pad < mPadsPerRow[row] - 2; ++pad) {
+      qMaxIndex = (timebin % 5) * mPadsPerRow[row] + pad;
```

Comment:
`const unsigned qMaxIndex = padOffset + pad;`
... this might reduce the number of calculations and memory access by caching the values.

Reply:
Yes, I already thought the compiler would take care.